File: Data Structures/Depth First Search/python/dfs.py
from collections import defaultdict

def DFS(v, visited):
    visited[v-1] = True
    print(v, end=" ")
    for u in graph[v]:
        if visited[u-1] == False:
            DFS(u, visited)

if __name__ == "__main__":
    #n = number of nodes, m = number of edges
    n, m = map(int, input().split())
    
    graph = defaultdict(list)

    #Input format: adjancey matrix of graph
    for _ in range(m):
        i, j = map(int, input().split())
        graph[i].append(j)
        graph[j].append(i)
    print(graph)

    visited = [False] * n

    DFS(1, visited)

#Analysis:

#Time Complexity: O(V + E)

File: Data Structures/Graphs/cpp/DFS.cpp
#include<iostream>
using namespace std;

void print( int** edges, int n, int sv, bool* visited){

  cout<<sv<<endl;
  visited[sv] = true;
  for( int i=0 ;i<n ; i++){
    if(i==sv)
    continue;
    if(edges[sv][i] == 1 && !visited[i]){
      print(edges, n, i, visited);
    }
  }
}

int main(){
  int n, e;
  cin>>n>>e;

  int** edges = new int*[n];

  for( int i=0; i<n; i++){
    edges[i] = new int[n];
    for(int j=0 ;j<n; j++){
      edges[i][j] = 0;
    }
  }

  bool* visited = new bool[n];

  for( int i=0; i<n; i++){
    visited[i] = false;
  }

  for( int i =0; i<e; i++){
    int f,s;
    cin>>f>>s;
    edges[f][s] = 1;
    edges[s][f] = 1;
  }

  print(edges, n, 0, visited);

  for( int i=0 ; i<n; i++){
    delete[] edges[i];
  }
  delete[] edges;

  delete[] visited;
}
